Updating Blake2 hash function to Blake3

GideonWhite1029 opened the pull request:
The code from this repository was used for the reference: https://github.com/bcgit/bc-java/blob/main/core/src/main/java/org/bouncycastle/crypto/digests/Blake3Digest.java

GideonWhite1029 committed:
Update to Blake3

Update hash function from Blake2 to Blake3

GideonWhite1029 committed:
Update to Blake3

Update hash function from Blake2 to Blake3

GideonWhite1029 committed:
Merge branch 'Winds-Studio:ver/1.21.8' into ver/1.21.8

HaHaWTH reviewed (commented):
Generally good, but need some benchmarks to make sure it's faster than original blake2b impl. Leave the old implementation as a compatibility option is better i assume

GideonWhite1029 committed:
Merge branch 'Winds-Studio:ver/1.21.8' into ver/1.21.8

GideonWhite1029 committed:
Optimize Blake3

- The execution time decreased from 1.822 ms to 757 ms
- Precomputed permutations have been added to compress()

GideonWhite1029 committed:
Optimize Blake3

- The execution time decreased from 1.822 ms to 757 ms
- Precomputed permutations have been added to compress()

GideonWhite1029 committed:
Merge branch 'Winds-Studio:ver/1.21.8' into ver/1.21.8

GideonWhite1029 committed:
Merge branch 'Winds-Studio:ver/1.21.8' into ver/1.21.8

GideonWhite1029 committed:
Added a choice of the type of hash in leaf-gloabl.yml

GideonWhite1029 committed:
Added a choice of the type of hash in leaf-gloabl.yml

Dreeam-qwq commented on leaf-server/src/main/java/su/plo/matter/WorldgenCryptoRandom.java:
```diff
@@ -24,6 +25,16 @@ public class WorldgenCryptoRandom extends WorldgenRandom {
     private final long[] message = new long[16];
     private final long[] cachedInternalState = new long[16];
 
+    private static long[] hashingWorldSeed(long[] seed) {
```
Should we move hashing algo type to config, and use interface-impl for 2b and 3?

cc @HaHaWTH

HaHaWTH replied:
Yes? That would be a cleaner implementation, anyway now looks good except this.

GideonWhite1029 committed:
Merge branch 'Winds-Studio:ver/1.21.8' into ver/1.21.8

GideonWhite1029 committed:
Merge branch 'Winds-Studio:ver/1.21.8' into ver/1.21.8

HaHaWTH reviewed (commented):
Will run some performance benchmarks and validations later

GideonWhite1029 committed:
Merge branch 'Winds-Studio:ver/1.21.8' into ver/1.21.8

GideonWhite1029 committed:
Merge branch 'Winds-Studio:ver/1.21.8' into ver/1.21.8

GideonWhite1029 committed:
Optimize Blake3 (again)
- The execution time decreased from 757 ms to 550 ms
- Changed logic of the type of hashing in WorldgenCryptoRandom

GideonWhite1029 committed:
Optimize Blake3 (again)
- The execution time decreased from 757 ms to 550 ms
- Changed logic of the type of hashing in WorldgenCryptoRandom

GideonWhite1029 committed:
Extending support for structures with Secure Seed
- Added hashing for structures: MineshaftStructure, OceanMonumentStructure, RuinedPortalStructure
- Small fix Balke3

GideonWhite1029 committed:
Extending support for structures with Secure Seed
- Added hashing for structures: MineshaftStructure, OceanMonumentStructure, RuinedPortalStructure
- Small fix Balke3

GideonWhite1029 commented:
@Dreeam-qwq Consider the current changes in the implementation of Blake3, as well as evaluate the purity of the code?

GideonWhite1029 committed:
Merge remote-tracking branch 'origin/ver/1.21.8' into ver/1.21.8

# Conflicts:
#	leaf-server/minecraft-patches/features/0132-Faster-random-generator.patch

GideonWhite1029 committed:
Fix apply patching

GideonWhite1029 committed:
Merge branch 'Winds-Studio:ver/1.21.8' into ver/1.21.8

GideonWhite1029 committed:
Merge branch 'Winds-Studio:ver/1.21.8' into ver/1.21.8

GideonWhite1029 committed:
Merge branch 'Winds-Studio:ver/1.21.8' into ver/1.21.8

GideonWhite1029 committed:
Merge branch 'Winds-Studio:ver/1.21.8' into ver/1.21.8

GideonWhite1029 committed:
Optimize Blake3 (again):
- Using SIMD instructions
- Each round is deployed to speed up hashing
- Execution time decreased from 550 ms to 340 ms

GideonWhite1029 committed:
Merge branch 'Winds-Studio:ver/1.21.8' into ver/1.21.8
